### PR TITLE
fix: Enable quantization and compilation in the same optimization job via ModelBuilder and add validations to prevent compilation for Llama-3.1 on TRTLLM.

### PR DESCRIPTION
**Pull request by cj-zhang**
…ModelBuilder.

*Issue #, if available:*

*Description of changes:* TRTLLM containers allow compilation and quantization at the same time, so the validations to ensure mutual exclusivity must be removed. Compilation using Llama-3.1 on TRTLLM v11 also doesn't currently work, so adding validations to prevent those jobs from being created.

*Testing done:* New and updated unit tests on jumpstart_builder and model_builder. Locally tested TRTLLM/Llama-3.1 validations from both JS and HF flow.

## Merge Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your pull request._

#### General

- [x] I have read the [CONTRIBUTING](https://github.com/aws/sagemaker-python-sdk/blob/master/CONTRIBUTING.md) doc
- [x] I certify that the changes I am introducing will be backward compatible, and I have discussed concerns about this, if any, with the Python SDK team
- [x] I used the commit message format described in [CONTRIBUTING](https://github.com/aws/sagemaker-python-sdk/blob/master/CONTRIBUTING.md#committing-your-change)
- [x] I have passed the region in to all S3 and STS clients that I've initialized as part of this change.
- [x] I have updated any necessary documentation, including [READMEs](https://github.com/aws/sagemaker-python-sdk/blob/master/README.rst) and [API docs](https://github.com/aws/sagemaker-python-sdk/tree/master/doc) (if appropriate)

#### Tests

- [x] I have added tests that prove my fix is effective or that my feature works (if appropriate)
- [x] I have added unit and/or integration tests as appropriate to ensure backward compatibility of the changes
- [x] I have checked that my tests are not configured for a specific region or account (if appropriate)
- [x] I have used [`unique_name_from_base`](https://github.com/aws/sagemaker-python-sdk/blob/master/src/sagemaker/utils.py#L77) to create resource names in integ tests (if appropriate)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/sagemaker/serve/builder/jumpstart_builder.py [hunk 1]
```diff
@@ -718,20 +718,23 @@ def _optimize_for_jumpstart(
...
         optimization_config, override_env = _extract_optimization_config_and_env(
             quantization_config, compilation_config
         )
-        if not optimization_config and is_compilation:
+        if (
+            not optimization_config or not optimization_config.get("ModelCompilationConfig")
+        ) and is_compilation:
             override_env = override_env or pysdk_model_env_vars
-            optimization_config = {
+            optimization_config["ModelCompilationConfig"] = {
```
Comment (GanAlps) [inline src/sagemaker/serve/builder/jumpstart_builder.py:737]: we are adding "ModelCompilationConfig" twice in this structure?

Comment (cj-zhang) [reply src/sagemaker/serve/builder/jumpstart_builder.py:745]: Whoops good catch. Fixed.

**Review by GanAlps — APPROVED** [review]

File: src/sagemaker/serve/builder/model_builder.py [hunk 2]
```diff
@@ -1279,6 +1276,28 @@ def _model_builder_optimize_wrapper(
             )
 
         if input_args:
+            optimization_instance_type = input_args["DeploymentInstanceType"]
+
+            # Compilation using TRTLLM and Llama-3.1 is currently not supported.
+            # TRTLLM is used by Neo if the following are provided:
+            #  1) a GPU instance type
+            #  2) compilation config
+            gpu_instance_families = ["g4", "g5", "p4d"]
```
Comment (xiongz945) [inline src/sagemaker/serve/builder/model_builder.py:1285]: Need to consider "g6", "p5" as well. And I don't "g4" should be included

File: tests/unit/sagemaker/serve/builder/test_js_builder.py [hunk 3]
```diff
@@ -1198,6 +1198,65 @@ def test_optimize_quantize_for_jumpstart(
...
+    def test_optimize_quantize_and_compile_for_jumpstart(
+        self,
+        mock_serve_settings,
+        mock_telemetry,
+    ):
+        mock_sagemaker_session = Mock()
+        mock_metadata_config = Mock()
+        mock_metadata_config.resolved_config = {
+            "supported_inference_instance_types": ["ml.inf2.48xlarge"],
+            "hosting_neuron_model_id": "huggingface-llmneuron-mistral-7b",
```
Comment (xiongz945) [inline tests/unit/sagemaker/serve/builder/test_js_builder.py:1212]: Why is this model ID not the same as the MB's init argument?

File: src/sagemaker/serve/builder/model_builder.py (hunk 2 above)
Comment (Lokiiiiii) [reply src/sagemaker/serve/builder/model_builder.py:1285]: g4 is not supported.

File: src/sagemaker/serve/builder/model_builder.py [hunk 4]
```diff
@@ -1279,6 +1276,28 @@ def _model_builder_optimize_wrapper(
...
+            is_gpu_instance = optimization_instance_type and any(
+                gpu_instance_family in optimization_instance_type
+                for gpu_instance_family in gpu_instance_families
+            )
+
+            # HF Model ID format = "meta-llama/Meta-Llama-3.1-8B"
+            # JS Model ID format = "meta-textgeneration-llama-3-1-8b"
+            llama_3_1_keywords = ["llama-3.1", "llama-3-1"]
```
Comment (Lokiiiiii) [inline src/sagemaker/serve/builder/model_builder.py:1293]: Can we swap this for a more dynamic check based on JS metadata ? 

I am not sure if this code path caters to non-JS models as well. So maybe we want to keep this for HF models.

File: tests/unit/sagemaker/serve/builder/test_js_builder.py [hunk 5]
```diff
@@ -1198,6 +1198,65 @@ def test_optimize_quantize_for_jumpstart(
...
+        model_builder.pysdk_model = mock_pysdk_model
+
+        out_put = model_builder._optimize_for_jumpstart(
+            accept_eula=True,
+            quantization_config={
+                "OverrideEnvironment": {"OPTION_QUANTIZE": "awq"},
+            },
+            compilation_config={"OverrideEnvironment": {"OPTION_TENSOR_PARALLEL_DEGREE": "2"}},
+            env_vars={
+                "OPTION_TENSOR_PARALLEL_DEGREE": "1",
```
Comment (Lokiiiiii) [inline tests/unit/sagemaker/serve/builder/test_js_builder.py:1252]: Why is there a mismatch in TP Degree here ?

Are we trying to compile with TP=2 and then deploy with TP=1 ?

This behavior is unsupported. Essentially LMI will ignore the runtime TP Degree in favor of what was compiled.

**Review by Lokiiiiii — CHANGES_REQUESTED** [review]
Missing items - 
1. The ability to add JS models as draft models for speculative decoding
2. If the draft model is a gated model, customer needs to provide an additional accept_eula parameter during model initialization.
3. TRT Compilation cannot be stacked with Speculative Decoding(SD)
4. New Quantization option - FP8
5. New Quantization option - SmoothQuant
6. SmoothQuant requires TRT compilation (llama 3.1 not supported, cannot be stacked with SD)
7. ModelSharding ...

**Comment by Lokiiiiii** [comment]
> Missing items -
> 
>     1. The ability to add JS models as draft models for speculative decoding
> 
>     2. If the draft model is a gated model, customer needs to provide an additional accept_eula parameter during model initialization.
> 
>     3. TRT Compilation cannot be stacked with Speculative Decoding(SD)
> 
>     4. New Quantization option - FP8
> 
>     5. New Quantization option - SmoothQuant
> 
>     6. SmoothQuant requires TRT compilation (llama 3.1 not supported, cannot be stacked with SD)
> 
>     7. ModelSharding ...

https://github.com/aws/sagemaker-python-sdk/blob/f45e6000d2afa6f12c544308a799fdcc194dfe7d/src/sagemaker/jumpstart/types.py#L1132 - ArtifactVersion is now optional and might not always be available in JS Metadata.

**Pull request closed without merging**